### PR TITLE
PE : parse Import, Delay_import tables & save imported_symbols

### PR DESCRIPTION
**Pull request by gougeon-s**
Before : - save dependencies from 'Import' table, not any imported_symbols
- One binary, many nodes (with --chroot) (fork version)
/my.dll
/sub-directory/my.dll
/home/user/chroot-dir/sub-directory/my.dll 
-Problem to build binmap.exe on Windows (CMakeList.txt)

Now : - save dependencies & imported_symbols from 'Import' & 'Delay_Import' tables
- One binary one node:
/sub-directory/my.dll (if present)
if not present : /./not-here.dll (easily to find them)
-Possible to build binmap.exe on Windows (delete one line on CMakeList.txt)

Could help to build on Windows : 
./binmap-master/_build> C:\Users\user\Documents\cmake-3.9.0-rc2-win64-x64\bin\cmake -
DBoost_DEBUG=ON -G "Visual Studio 12" -DBoost_USE_STATIC_LIBS=ON -
DBOOST_ROOT=C:\local\boost_1_56_0 -DBOOST_LIBRARYDIR=C:\local\boost_1_56_0\lib32-msvc-
12.0 -DZLIB_LIBRARY="C:\Program Files (x86)\zlib" -DZLIB_INCLUDE_DIR="C:\Program Files
(x86)\zlib\include" -DPYTHON_INCLUDE_DIR="C:\Python27\include" -
DPYTHON_LIBRARY="C:\Python27" ..

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
wow, the windows integration part looks ok (could be a seperate PR to make everything easier), can you detail the feature implemented by the delay import mechanism? i'll review that tomorrow :-)

**Comment by gougeon-s** [comment]
There are two import tables in PE binaires : Import and Delay_import, I've made the structure of Delay_Import in PE.hpp. It parses the two tables to get the deps, then parses each ones again to get the symbols (decoder.cpp)

Actually, I've made all the changes in the same version. So, what are the PR you'd like? It may be possible to do it.

**Comment by serge-sans-paille** [comment]
It's ok (and actually pretty nice) to have the delay_import feature in a single commit. if you could just extract the small change related to the Windows build and put it in a seperate PR, that would be great. I'll start to review your code afterward.

File: src/graph.cpp [hunk 1]
```diff
@@ -135,17 +145,67 @@ void Graph::compute_distance_matrix() const {
...
+
+
+    boost::filesystem::path no_path1 = "/.";
+    boost::filesystem::path no_path2 = ".";
+    std::string file = input_file.string();
+
+    if(input_file.parent_path() == no_path1 || input_file.parent_path() == no_path2){
+	//std::cout<<"/. or .\n";
+
+	for (it_p_=passed_path_.begin(); it_p_!=passed_path_.end();it_p_++){
```
Comment (serge-sans-paille) [inline src/graph.cpp]: I don't get why this ``it_p_`` is a member variable, it seems to be used as a local one only.

File: src/graph.cpp [hunk 2]
```diff
@@ -135,17 +145,67 @@ void Graph::compute_distance_matrix() const {
...
+		if(it_p_->first != no_path1 && 
+		   it_p_->first != no_path2 &&
+		   has_node(it_p_->first / input_file.filename())){
+			return (it_p_->first / input_file.filename());
+		}
+	}
+    }else{
+	//std::cout << "neither /. or .\n";
```
Comment (serge-sans-paille) [inline src/graph.cpp]: there's a log of debug code / comment, if you could clean that too.

File: src/graph.cpp [hunk 3]
```diff
@@ -135,17 +145,67 @@ void Graph::compute_distance_matrix() const {
...
+    std::string file = input_file.string();
+
+    if(input_file.parent_path() == no_path1 || input_file.parent_path() == no_path2){
+	//std::cout<<"/. or .\n";
+
+	for (it_p_=passed_path_.begin(); it_p_!=passed_path_.end();it_p_++){
+		if(it_p_->first != no_path1 && 
+		   it_p_->first != no_path2 &&
+		   has_node(it_p_->first / input_file.filename())){
+			return (it_p_->first / input_file.filename());
```
Comment (serge-sans-paille) [inline src/graph.cpp]: So that's why you changed the add_node return type. Can you explain what's going on?

File: src/scan.cpp [hunk 4]
```diff
@@ -175,8 +179,8 @@ class Scanner {
 
   void analyze_metadata(Collector &collector,
                         Hash const &input_hash) {
-    try {
-      MetadataInfo mi(input_hash);
```
Comment (serge-sans-paille) [inline src/scan.cpp]: if you could avoid this reformating too, they don't help for the review.

File: src/scan.cpp [hunk 5]
```diff
@@ -107,27 +107,31 @@ class Scanner {
...
-                                           << std::endl;
-            // still write dummy metadata for consistency
-            blobmap_.metadata()->insert(MetadataInfo(input_hash));
-          }
+          boost::filesystem::path path_to_add = add_node(input_file, input_hash);
+
+	  if (path_to_add == input_file || visited_.find(path_to_add) == visited_.end()) {
+		visited_.insert(path_to_add);
+
+		std::set<boost::filesystem::path> deps;
```
Comment (serge-sans-paille) [inline src/scan.cpp]: any reasoj why this set is not local to the try block?

File: src/graph.cpp (hunk 3 above)
Comment (gougeon-s) [inline src/graph.cpp]: For example, without this, if we 're scanning /sub-dir/one.dll depending to two.dll in the chroot (/two.dll) it won't find it and save it as it isn't here (/./two.dll). With this, if we've parsed /two.dll, it will find it and continues with this path.

Actually this function do this schema : 
/./dep.dll (not in the path) => do we have parsed it in an other directory?
   yes => save /the-dir/dep.dll
   no=> save /./dep.dll

/dir/dep2.dll found : 
    /./dep2.dll already saved ?
       yes : changes to /dir/dep2.dll
       no : save /dir/dep2.dll

File: src/scan.cpp (hunk 5 above)
Comment (gougeon-s) [inline src/scan.cpp]: deps? It was already like this (master, line 112)

File: src/scan.cpp (hunk 4 above)
Comment (gougeon-s) [inline src/scan.cpp]: I agree, I will avoid next time

File: src/graph.cpp (hunk 1 above)
Comment (gougeon-s) [inline src/graph.cpp]: I think it's local now, tell me if I am wrong

File: src/graph.cpp (hunk 3 above)
Comment (serge-sans-paille) [reply src/graph.cpp]: Interesting. I still have a question: even if a library with the same name was already parsed in another directory, why would we link with it? If the library is not in a standard path, there's no reason why we should resolve to that one.

As an exemple, there may be *two* libraries with the same name in two different directories. In that case, to which should we resolve?

File: include/binmap/graph.hpp [hunk 6]
```diff
@@ -55,6 +55,10 @@ class Graph {
   mapping_;
   mutable std::vector<int> *distance_matrix_;
 
+  boost::unordered_map<boost::filesystem::path, int> passed_path_;
+  boost::unordered_map<boost::filesystem::path, graph_type::vertex_descriptor>::iterator
+	map_p_;
```
Comment (serge-sans-paille) [inline include/binmap/graph.hpp]: this `map_p_` member is not used anymore

File: include/binmap/graph.hpp [hunk 7]
```diff
@@ -55,6 +55,10 @@ class Graph {
   mapping_;
   mutable std::vector<int> *distance_matrix_;
 
+  boost::unordered_map<boost::filesystem::path, int> passed_path_;
```
Comment (serge-sans-paille) [inline include/binmap/graph.hpp]: From its usage, it seems to me this could be a set and not a map. Also, it's very similar to the visited variable in `scan.cpp`, may be you could call it like this too? or `visited_path`.

File: src/graph.cpp (hunk 3 above)
Comment (serge-sans-paille) [reply src/graph.cpp]: Once this is settled, you could return a reference to the path stored in the graph instead of making a copy.

File: src/scan.cpp [hunk 8]
```diff
@@ -209,15 +215,22 @@ class Scanner {
...
+    //TODO: fix for windows native() vs. string()
     std::string const& root = Env::root().string();
+    std::string dll_name = path.filename().string();
+
+    boost::algorithm::to_lower(dll_name);
+    boost::filesystem::path path_low = (path.parent_path() / dll_name);
+    
+    std::string const& spath = path_low.string();
     // only trim root if root is a prefix
+    //std::cout << "\nspath: " << spath << " root: " << root << " path " << path << "\n";
```
Comment (serge-sans-paille) [inline src/scan.cpp]: remaining comment.

File: src/scan.cpp [hunk 9]
```diff
@@ -185,14 +189,16 @@ class Scanner {
     }
   }
 
-  void
+  boost::filesystem::path
   add_node(boost::filesystem::path const &input_file,
                             Hash const &input_hash) {
+    //logging::log(logging::warning) << "\nadding file: " << input_file<<"\n";
```
Comment (serge-sans-paille) [inline src/scan.cpp]: remaining comment.

File: src/collectors/pe/decoder.cpp [hunk 10]
```diff
@@ -537,6 +610,280 @@ bool pe_decoder<_Bits>::find_module_path(boost::filesystem::path const &containi
...
+struct PeImageDelayImport {
+	uint32_t dd_grAttrs;
+	uint32_t szName;
+	uint32_t phmod;
+	uint32_t pIAT;
+	uint32_t pINT;
+	uint32_t pBoundIAT;
+	uint32_t pUnloadIAT;
+	uint32_t dwTimeStamp;
+};
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: remaining comments

File: src/collectors/pe/decoder.cpp [hunk 11]
```diff
@@ -537,6 +610,280 @@ bool pe_decoder<_Bits>::find_module_path(boost::filesystem::path const &containi
...
+		for (int k=0; k<step; k++){
+			func_rva += (uint64_t)(func_buff[k]+256)%256 * (int)(pow(256,k));
+		}
+
+		if(func_rva ==0){
+			logging::log(logging::info) << "func_rva=0\n";
+			break;
+		}
+		
+		//printf("step: %d\nfunc_rva: %08x, %llx\n", step, func_rva, func_rva);
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: remaining comments

**Comment by serge-sans-paille** [comment]
@gougeon-s I di not review everything, I'd like to settle the ``add_node`` returns a path before.

File: src/graph.cpp (hunk 3 above)
Comment (gougeon-s) [reply src/graph.cpp]: Actually, if there were (without this) two libraries in different paths, there will be 4 or 5 nodes : 
/lib.dll (not found)
/dir1/lib.dll (first lib parsed)
/dir2/lib.dll (2nd lib parsed)
/home/user/chroot-dir/dir1/lib.dll    or/and    /home/user/chroot-dir/dir2/lib.dll depending the lib in the sub-directories 
 
I had to make a choice, so now there can be 2 different nodes /dir1/lib.dll and /dir2/lib.dll. The most probable reason to have this happening : - scan an entire system so it can mostly be improved if we store the difference 64bits/32bits and link only if it is the same. - chroot folder hand-made and we have stored twice the same dll in different sub-directories.

So, it can be improved, but I think it's better to have 2 nodes if they are in different paths than 4 or 5

Comment (gougeon-s) [reply src/graph.cpp]: For the reference, if we don't return (it_p_->first / input_file.filename()), then we have to find mapping_[(it_p_->first / input_file.filename())] then extract the path/filename from his descriptor. because after it use only this. That's a loop (form path/filename to path/filename)we can avoid I think, don't you? (or I didn't get what you meant)

File: include/binmap/graph.hpp (hunk 7 above)
Comment (gougeon-s) [reply include/binmap/graph.hpp]: You're right, that will be more coherent

File: src/binmap.cpp [hunk 12]
```diff
@@ -155,7 +155,7 @@ class SubCommand : public Command {
   }
 };
 
-struct canonicalize_path {
+struct canonicalize {
```
Comment (serge-sans-paille) [inline src/binmap.cpp]: why this change?

File: src/collectors/pe/decoder.cpp [hunk 13]
```diff
@@ -257,7 +258,7 @@ void pe_decoder<_Bits>::get_imports(boost::filesystem::path const &module_path,
         else{
             // not a redirection scheme and no import table...
             // Note: do NOT throw, we might have modules that do not export anything!
-            logging::log(logging::error) << "module " << module_path << " has no import table." << std::endl;
+            logging::log(logging::error) << "ERROR : module " << module_path << " has no import table." << std::endl;
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: if you really want to change the dispaly of errors, then it should be done at the logger level, not at the call site.

File: src/collectors/pe/decoder.cpp [hunk 14]
```diff
@@ -296,33 +297,105 @@ void pe_decoder<_Bits>::get_imports(boost::filesystem::path const &module_path,
...
+
+	std::transform(dep_name.begin(), dep_name.end(), dep_name.begin(), ::tolower);
+
+
+	boost::filesystem::path full_path;
+        // try to resolve full module path
+        if (!find_module_path(module_path, dep_name)){
+            logging::log(logging::warning) << "couldn't find full path for module " << dep_name
+                << " imported by module " << module_path << std::endl;
+		full_path = Env::root() / "./" / dep_name;
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: this single ``"./" / `` breaks our validation. Why is it there?

**Comment by serge-sans-paille** [comment]
@gougeon-s I impelmnted the error formating stuff in a seperate commit to make it easier to keep this PR simple, see d92e35c739be10df927aa3db6d5e0064f4717662

File: src/binmap.cpp (hunk 12 above)
Comment (gougeon-s) [reply src/binmap.cpp]: Ah, that's a change made in the last commit, I worked on the previous one, I can change this

File: src/collectors/pe/decoder.cpp (hunk 14 above)
Comment (gougeon-s) [reply src/collectors/pe/decoder.cpp]: That's only for the lib not found. If it's found later in an other path, it's changed. That's why at the end, we can find the lib not reached './lib.dll'

**Comment by gougeon-s** [comment]
Nice. Well, I'm sorry for the many annoying things (comments etc) or if I misused something. It is the first time I contribute to a project.

**Comment by serge-sans-paille** [comment]
That's OK, we're heading in the right direction, and the actual content of the PR is great, just have to sort out the noise :-)

File: src/collectors/pe/decoder.cpp (hunk 14 above)
Comment (serge-sans-paille) [reply src/collectors/pe/decoder.cpp]: ok, why not. But in the end I have this path in the data abse: '/./truc.dll' and that's not canonical...

File: include/binmap/graph.hpp [hunk 15]
```diff
@@ -72,7 +74,7 @@ class Graph {
   bool has_path(boost::filesystem::path const &from,
                 boost::filesystem::path const &to) const;
 
-  void add_node(boost::filesystem::path const &input_file,
+  boost::filesystem::path add_node(boost::filesystem::path const &input_file,
```
Comment (serge-sans-paille) [inline include/binmap/graph.hpp]: Please document the behavior with respect to add_node return value here

File: src/collectors/pe/data.cpp [hunk 16]
```diff
@@ -123,6 +123,27 @@ bool PeData<_Bits>::convert_rva_to_offset(uint32_t rva,
     return false;
 }
 
+
+template <typename _Bits>
+bool PeData<_Bits>::convert_rva_to_offset64(uint64_t rva,
+    uint64_t &offset) const {
+
+    if (_sections.empty()){
```
Comment (serge-sans-paille) [inline src/collectors/pe/data.cpp]: this test is useless, if _sections is empty, then the loop below is never run and it still returns false.

File: src/collectors/pe/decoder.cpp [hunk 17]
```diff
@@ -310,19 +311,91 @@ void pe_decoder<_Bits>::get_imports(boost::filesystem::path const &module_path,
...
         std::transform(dll_name.begin(), dll_name.end(), dll_name.begin(), ::tolower);
 
+	boost::filesystem::path full_path;
         // try to resolve full module path
         if (!find_module_path(module_path, dll_name)){
             logging::log(logging::warning) << "couldn't find full path for module " << dll_name
                 << " imported by module " << module_path << std::endl;
-        }
+		full_path = Env::root() / "./" / dll_name;
+		logging::log(logging::warning) << "dll name:" << full_path << "\n";
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: strange / incosistent indentation

File: src/collectors/pe/decoder.cpp [hunk 18]
```diff
@@ -310,19 +311,91 @@ void pe_decoder<_Bits>::get_imports(boost::filesystem::path const &module_path,
...
         imports.insert(full_path);
 
+
     } while (::memcmp(&imp_desc, &imp_end, sizeof(PeImportDescriptor)));
+
+
+	uint32_t delay_imports_rva =
+        _pe_data.nt_headers().OptionalHeader.DataDirectory[PeDataDirectory::kEntryDelayImport]
+        .VirtualAddress;
+    uint32_t delay_imports_len =
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: again, strange indent

File: src/collectors/pe/decoder.cpp [hunk 19]
```diff
@@ -537,6 +609,248 @@ bool pe_decoder<_Bits>::find_module_path(boost::filesystem::path const &containi
...
+    {
+        //check if this module is part of the apisetschema redirection scheme
+        //    in this case there's no import table, we'll do the redirection statically.
+        std::string module_filename = module_path.filename().string();
+        if (boost::starts_with(module_filename, WINDOWS_APISETSCHEMA_API_START) ||
+            boost::starts_with(module_filename, WINDOWS_APISETSCHEMA_EXT_START))
+        {
+            std::string empty;
+            find_module_path(module_path, empty);
+	    logging::log(logging::warning) << "Apisetschema redirection scheme\n";
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: here too :-)

Comment (serge-sans-paille) [reply src/collectors/pe/decoder.cpp]: and you're using '\n' instead of std::endl

File: src/collectors/pe/decoder.cpp [hunk 20]
```diff
@@ -537,6 +609,248 @@ bool pe_decoder<_Bits>::find_module_path(boost::filesystem::path const &containi
...
+
+    } while (::memcmp(&imp_desc, &imp_end, sizeof(PeImportDescriptor)));
+	logging::log(logging::info) <<"out_imported_symbols\n";
+	return true;
+}
+
+template <typename _Bits>
+bool pe_decoder<_Bits>::get_delay_imports(const boost::filesystem::path  &module_path, std::vector<std::string> &imported_symbols) const
+{
+	logging::log(logging::info) <<"start_delay_imports\n";
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: endl;

File: src/collectors/pe/decoder.cpp [hunk 21]
```diff
@@ -537,6 +609,248 @@ bool pe_decoder<_Bits>::find_module_path(boost::filesystem::path const &containi
...
+    uint32_t delay_imports_rva =
+        _pe_data.nt_headers().OptionalHeader.DataDirectory[PeDataDirectory::kEntryDelayImport]
+        .VirtualAddress;
+    uint32_t delay_imports_len =
+        _pe_data.nt_headers().OptionalHeader.DataDirectory[PeDataDirectory::kEntryDelayImport]
+        .Size;
+
+    if (delay_imports_rva == 0 || delay_imports_len == 0){
+	logging::log(logging::info) <<"no_delay_import: out\n";
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: endl;

File: src/graph.cpp [hunk 22]
```diff
@@ -135,17 +144,57 @@ void Graph::compute_distance_matrix() const {
...
-void Graph::add_node(boost::filesystem::path const &input_file,
+boost::filesystem::path Graph::add_node(boost::filesystem::path const &input_file,
                      Hash const &input_hash) {
-  assert(not has_node(input_file));
-#pragma omp critical(graph_)
-  {
+  if(not has_node(input_file)){
+
+    if (visited_path_.find(input_file.parent_path()) == visited_path_.end()){ //path not parsed yet
+	visited_path_[input_file.parent_path()] = 1;
```
Comment (serge-sans-paille) [inline src/graph.cpp]: Again, visited_path sould be a (unrodered ?) set, not a map

File: src/scan.cpp [hunk 23]
```diff
@@ -185,14 +189,15 @@ class Scanner {
...
+  boost::filesystem::path
   add_node(boost::filesystem::path const &input_file,
                             Hash const &input_hash) {
     boost::filesystem::path const trimed_input_file = trim_root(input_file);
     Graph &graph = current_graph();
-    graph.add_node(trimed_input_file, input_hash);
-    logging::log(logging::info) << "adding file: " << trimed_input_file << " "
-                                << input_hash << std::endl;
+    boost::filesystem::path to_add = graph.add_node(trimed_input_file, input_hash);
+    logging::log(logging::warning) << "\nadding file: " << to_add << " "
```
Comment (serge-sans-paille) [inline src/scan.cpp]: the "\n" prefix inj the log is a bad idea

**Comment by serge-sans-paille** [comment]
@gougeon-s I've made my final remarks, once everything is update, you can rebase on master, squash all your changes in one commit and that will be ok. Thanks again for the contribution and your patience during the review process.

File: include/binmap/graph.hpp (hunk 15 above)
Comment (gougeon-s) [reply include/binmap/graph.hpp]: graph.add_node return a path which is used in scan.add_node (this one return the same path.)

It's then used in operator() : 
if the return of add_node was the same as input_file (that means it hasn't been parsed before and it is in the same path as the binary who had the dependency or it isn't in any path found before) we continue,

if the return of add_node gives an other binary : we look if it hasn't been parsed before. If it hasn't we continue and add to the visited_ (actually we add in both cases, because there isn't problem if we have the same bin twice in visited_ , but we can change it.)

After, we continue the analyze with the binary found. (analyze_dependencies, analyze_metadata, add_deps)

**Comment by gougeon-s** [comment]
I have done the changes needed. I hope there won't have any issue with the indent (some tabs have been changed into spaces..?)
If it is ok with this commit, I will do a new & cleaner PR  tomorrow

File: include/binmap/graph.hpp (hunk 15 above)
Comment (serge-sans-paille) [reply include/binmap/graph.hpp]: ok, I meant to add this as a comment in the code, something like

```c++
/**
 * if `input_file` was not met before, add it to the graph and return it. Otherwise ...
 */
```

File: src/collectors/pe/decoder.cpp [hunk 24]
```diff
@@ -310,19 +311,91 @@ void pe_decoder<_Bits>::get_imports(boost::filesystem::path const &module_path,
...
 
+		boost::filesystem::path full_path;
         // try to resolve full module path
         if (!find_module_path(module_path, dll_name)){
-            logging::log(logging::warning) << "couldn't find full path for module " << dll_name
+			logging::log(logging::warning) << "couldn't find full path for module " << dll_name
                 << " imported by module " << module_path << std::endl;
-        }
+			full_path = Env::root() / "./" / dll_name;
+			logging::log(logging::warning) << "dll name:" << full_path <<  std::endl;
```
Comment (serge-sans-paille) [inline src/collectors/pe/decoder.cpp]: sotty about that, but looks like you're spacing is configured strangely. Mixing tab and spaces?

**Comment by serge-sans-paille** [comment]
Appart from the tabbing and one missing doc point *in the code* I'm ok with the PR, thanks *a lot*

File: src/collectors/pe/decoder.cpp (hunk 24 above)
Comment (gougeon-s) [reply src/collectors/pe/decoder.cpp]: "sed 's/ /\t/g' decoder.cpp > decoder.cpp" has done the job, but github will think that I have change the entire file

**Comment by gougeon-s** [comment]
Thank you for all the advices, I have made a new PR with only 1 commit :)

**Comment by serge-sans-paille** [comment]
closed by #14
